src/routing/phase.cc

```diff
@@ -0,0 +1,171 @@
+                                   [](const auto& pos) { return pos.lane != nullptr; }));
+  MALIPUT_THROW_UNLESS(!end_positions_.empty());
+  MALIPUT_THROW_UNLESS(
+      std::all_of(end_positions_.begin(), end_positions_.end(), [](const auto& pos) { return pos.lane != nullptr; }));
+  MALIPUT_THROW_UNLESS(!lane_s_ranges_.empty());
+  MALIPUT_THROW_UNLESS(road_network_ != nullptr);
+  const api::RoadGeometry* road_geometry = road_network_->road_geometry();
+  MALIPUT_VALIDATE(std::all_of(lane_s_ranges_.begin(), lane_s_ranges_.end(),
+                               [&](const auto& lane_s_range) {
+                                 return ValidateLaneSRangeIsInRoadNetwork(lane_s_range, road_geometry,
+                                                                          lane_s_range_tolerance_);
+                               }),
+                   "LaneSRange is not in the road_network_.");
+  MALIPUT_VALIDATE(std::all_of(start_positions_.begin(), start_positions_.end(),
+                               [&](const auto& position) {
+                                 return ValidatePositionIsInLaneSRanges(position, lane_s_ranges_,
+                                                                        lane_s_range_tolerance_);
+                               }),
+                   "Start position is not in lane_s_ranges_.");
+  MALIPUT_VALIDATE(std::all_of(end_positions_.begin(), end_positions_.end(),
+                               [&](const auto& position) {
+                                 return ValidatePositionIsInLaneSRanges(position, lane_s_ranges_,
+                                                                        lane_s_range_tolerance_);
+                               }),
+                   "End position is not in lane_s_ranges_.");
+  for (size_t i = 0; i < lane_s_ranges_.size() - 1u; ++i) {
+    MALIPUT_VALIDATE(ValidateLaneSRangesAreAdjancent(lane_s_ranges_[i], lane_s_ranges_[i + 1], road_geometry),
+                     "LaneSRanges are not adjacent.");
+  }
+}
+
+PhasePositionResult Phase::FindPhasePositionBy(const api::InertialPosition& inertial_position) const {
+  std::vector<api::RoadPositionResult> road_position_results;
+  const api::RoadGeometry* road_geometry = road_network_->road_geometry();
+  std::transform(lane_s_ranges_.cbegin(), lane_s_ranges_.cend(), std::back_inserter(road_position_results),
+                 [road_geometry, &inertial_position](const auto& lane_s_range) {
+                   const api::Lane* lane = road_geometry->ById().GetLane(lane_s_range.lane_id());
+                   const api::LanePositionResult lane_position_result = lane->ToLanePosition(inertial_position);
+                   return api::RoadPositionResult{api::RoadPosition{lane, lane_position_result.lane_position},
+                                                  lane_position_result.nearest_position, lane_position_result.distance};
+                 });
```

Comment:
Is it expected the lane_s_ranges to contain ranges that are smaller portions than the length of the lane?
If so we should ensure that the road position result doesn't point to an `s` out of the range.